tablets: reload only changed metadata

denesb opened the pull request:
Currently, each change to tablet metadata triggers a full metadata reload from disk. This is very wasteful, especially if the metadata change affects only a single row in the `system.tablets` table. This is the case when the tablet load balancer triggers a migration, this will affect a single row in the table, but today will trigger a full reload.
We expect tablet count to potentially grow to thousands and beyond and the overhead of this full reload can become significant.
This PR makes tablet metadata reload partial, instead of reloading all metadata on topology or schema changes, reload only the partitions that are affected by the change. Copy the rest from the in-memory state.
This is done with two passes: first the change mutations are scanned and a hint is produced. This hint is then passed down to the reload code, which will use it to only reload parts (rows/partitions) of the metadata that has actually changed.

The performance difference between full reload and partial reload is quite drastic:
```
INFO  2024-07-25 05:06:27,347 [shard 0:stat] testlog - Tablet metadata reload:
full      616.39ms
partial     0.18ms
```
This was measured with the modified (by this PR) `perf_tablets`, which creates 100 tables, each with 2K tablets. The test was modified to change a single tablet, then do a full and partial reload respectively, measuring the time it takes for reach.

Fixes: #15294

New feature, no backport needed.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3915/

denesb commented:
v2:
* Implement partial reload for schema changes too.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3930/

avikivity commented:
What's the goal? reduce CPU cycles?

tgrabiec commented on locator/tablets.cc:
```diff
@@ -132,6 +132,12 @@ future<> tablet_metadata::clear_gently() {
 }
 
+void tablet_metadata::merge(tablet_metadata&& other) {
+    for (auto&& [id, map] : other._tablets) {
```
I think this doesn't work with deletions. If there is a hint, but element is absent in other, we should delete.

tgrabiec commented:
> What's the goal? reduce CPU cycles?

Yes. It's a step towards avoiding quadratic behavior when applying tablet metadata changes. Similar to what we did for schema changes in c8e2bf1596be35f1a81004ba145a7e51ec748907.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:x: - Unit Tests
#### Failed Tests (36/21011):
- [topology_experimental_raft.test_tablets.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4380/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_tablets_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.debug.1)
- [topology_experimental_raft.test_tablets.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4380/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_tablets_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.debug.2)
- [topology_experimental_raft.test_tablets.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4380/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_tablets_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.debug.3)
- [topology_experimental_raft.test_tablets.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4380/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_tablets_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.dev.1)
- [topology_experimental_raft.test_tablets.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4380/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_tablets_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.dev.2)
- [topology_experimental_raft.test_tablets.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4380/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_tablets_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.dev.3)
- [topology_experimental_raft.test_tablets.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4380/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_tablets_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.release.1)
- [topology_experimental_raft.test_tablets.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4380/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_tablets_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.release.2)
- [topology_experimental_raft.test_tablets.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4380/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_tablets_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.release.3)
- [test_tablet_metadata_propagates_with_schema_changes_in_snapshot_mode](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4380/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_tablet_metadata_propagates_with_schema_changes_in_snapshot_mode) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_metadata_propagates_with_schema_changes_in_snapshot_mode)
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4380/
- Duration: 2 hr 29 min
- Builder: spider7.cloudius-systems.com

tgrabiec commented:
> v3:
> 
> * Instead of building partial `tablet_metadata` from hints, then merging with previous value, pass down the previous value and update in-place.
> * Add tests
> 
> @tgrabiec I need some help. I have added tests and so far I could not find a scenario which would require either row-level updates or even deletions. I cloned `test_tablet_metadata_persistence` (see last commit) and made it do partial updates to test my changes. Can you give me some example operations that would produce partition deletion and row-level changes? I mean I could build mutations that do this, but I want to make sure I'm testing with mutations that look like those produced by real tablet updates, instead of something fully synthetic.

Partition deletions are produced by a table drop, here in tablet_allocator.cc:

```
    void on_before_drop_column_family(const schema& s, std::vector<mutation>& muts, api::timestamp_type ts) override {
        keyspace& ks = _db.find_keyspace(s.ks_name());
        auto&& rs = ks.get_replication_strategy();
        std::vector<mutation> result;
        if (rs.uses_tablets()) {
            auto tm = _db.get_shared_token_metadata().get();
            muts.emplace_back(make_drop_tablet_map_mutation(s.keypace_name(), s.id(), ts));
        }
    }
``` 

Row updates are produced by the topology change coordinator based on decision given by load balancer:

```c++
                slogger.trace("raft topology: Will set tablet {} stage to {}", gid, stage);
                updates.emplace_back(
                    replica::tablet_mutation_builder(guard.write_timestamp(), s->ks_name(), table)
                        .set_stage(last_token, stage)
                        .build());

``` 

There is no way to explicitly trigger tablet migration yet, but you can enable shuffle mode via error injection on `tablet_allocator_shuffle` injection point to make sure there are migrations constantly. Alternatively, you can build a mutation manually and execute it in group0.

test_tablets.py should cover such migrations already.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4837/
- Duration: 2 hr 37 min
- Builder: spider2.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:x: - Unit Tests
#### Failed Tests (1/21877):
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4850/testReport/junit/boost.view_complex_test.test_marker_timestamp_is_not_shadowed_by_previous_updatewith_flush.3.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4850/
- Duration: 2 hr 28 min
- Builder: spider5.cloudius-systems.com

denesb commented:
v5:
* Add tablet metadata reload perf measurement to `test_large_tablet_metadata`.

denesb commented:
> #### 🔴 CI State: FAILURE
> 
> ✅ - Build ✅ - Sanity Tests ❌ - Unit Tests
> #### Failed Tests (1/21877):
> 
>     * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4850/testReport/junit/boost.view_complex_test.test_marker_timestamp_is_not_shadowed_by_previous_updatewith_flush.3.xunit/xml/Tests___Unit_Tests____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+%5Bempty%5D)
> 
> 
> #### Build Details:
> 
>     * Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4850/
> 
>     * Duration: 2 hr 28 min
> 
>     * Builder: spider5.cloudius-systems.com

https://github.com/scylladb/scylladb/issues/16144

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:white_check_mark: - Build
:x: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4853/
- Duration: 2 hr 24 min
- Builder: spider4.cloudius-systems.com

denesb commented:
> #### 🔴 CI State: ABORTED
> 
> ✅ - Build ❌ - Sanity Tests ✅ - Unit Tests
> #### Build Details:
> 
>     * Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4853/
> 
>     * Duration: 2 hr 24 min
> 
>     * Builder: spider4.cloudius-systems.com

Spot termination.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4865/
- Duration: 2 hr 14 min
- Builder: spider5.cloudius-systems.com

denesb commented:
@tgrabiec ping.

denesb commented:
@tgrabiec ping^2.

mykaul commented:
@denesb - need to resolve conflicts here.

tgrabiec commented on test/boost/tablets_test.cc:
```diff
@@ -796,11 +802,52 @@ SEASTAR_TEST_CASE(test_large_tablet_metadata) {
+
+        BOOST_REQUIRE_EQUAL(tm, tm_full_reload);
+
+        testlog.info("Tablet metadata reload:\nfull    {:>8.2f}ms\npartial {:>8.2f}ms", full_reload_duration.count(), partial_reload_duration.count());
```
Can we move this to perf_tablets?

denesb replied:
Yes, fixed in v5 (also added tests for this case).

tgrabiec commented on locator/tablets.cc:
```diff
@@ -109,28 +109,91 @@ tablet_replica get_leaving_replica(const tablet_info& tinfo, const tablet_transi
-tablet_map& tablet_metadata::get_tablet_map(table_id id) {
-    return const_cast<tablet_map&>(
-        const_cast<const tablet_metadata*>(this)->get_tablet_map(id));
+void tablet_metadata::mutate_tablet_map(table_id id, noncopyable_function<void(tablet_map&)> func) {
```
Since tablet_metadata is part of token metadata, its updates should be already isolated by the means of mutate_token_metadata(), and work on a temporary copy before it's published to all shards.
Do we really need mutate_tablet_map?

denesb replied:
Strictly speaking yes, but I still wanted to enforce immutability on the API level.

tgrabiec commented on replica/tablets.cc:
```diff
@@ -156,6 +156,65 @@ future<> save_tablet_metadata(replica::database& db, const tablet_metadata& tm,
+    if (mp.partition_tombstone() || !mp.row_tombstones().empty()) {
+        // If there is a partition tombstone or range tombstone, update the
+        // entire partition.
+        return;
```
So empty "tokens" means that whole table is affected. But what if we already had a hint for some tokens, but now we decide that whole table is affected. Shouldn't we clear "tokens" to indicate that?

denesb replied:
Fixed in v5 (also added tests for this case).

tgrabiec commented on replica/tablets.cc:
```diff
@@ -156,6 +156,65 @@ future<> save_tablet_metadata(replica::database& db, const tablet_metadata& tm,
+        // entire partition.
+        return;
+    }
+
```
Need to also handle static column changes.

denesb replied:
Fixed in v5 (also added tests for this case).

bhalevy commented:
Deferring to 6.1

mykaul commented:
I assume this should be deferred to 6.2?

denesb commented:
> I assume this should be deferred to 6.2?

Yes, no way this makes it into 6.1. I successfully rebased it but almost al the tests are failing, the investigations can take an uncertain amount of time.

denesb commented:
v5:
* rebased
* moved reload perf test to perf_tablets
* added specific tests for change hint update, in particular for partition, range and row tombstones and static rows
* fixed handling of partition, range and row tombstones and static rows

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10724/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10724/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/schema_change_test
:small_blue_diamond: boost/tablets_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10724/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2888/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10910/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10724/)
#### Build Details:
- Duration: 3 hr 37 min
- Builder: spider5.cloudius-systems.com

avikivity commented:
Missing: label paperwork (labelwork?)

denesb commented:
> Missing: label paperwork (labelwork?)

Yes, this PR precedes all the new label bureaucracy, fixed now.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10857/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10857/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/schema_change_test
:small_blue_diamond: boost/tablets_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10857/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10857/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11017/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2994/)
#### Build Details:
- Duration: 3 hr 11 min
- Builder: spider7.cloudius-systems.com

denesb commented on db/schema_tables.cc:
```diff
@@ -1554,13 +1553,13 @@ static future<> merge_tables_and_views(distributed<service::storage_proxy>& prox
         // We must also do it before tables are created so that new tables see the tablet map.
         co_await db.invoke_on_all([&] (replica::database& db) -> future<> {
-            co_await db.get_notifier().update_tablet_metadata();
+            co_await db.get_notifier().update_tablet_metadata(std::move(tablet_hint));
```
Please send a PR to remove it, you found the bug, you should get the credit for fixing it. ;)

tgrabiec replied:
It's already part of a pending PR: https://github.com/scylladb/scylladb/pull/20853/commits/3119a02eddd0df45c4c0cae22c8a8d1d86f17b20
